File: app/src/main/java/ir/arapp/arappmain/View/Fragments/RegisterPhoneFragment.java
package ir.arapp.arappmain.View.Fragments;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import java.util.Objects;
import ir.arapp.arappmain.R;
import ir.arapp.arappmain.Util.Services.FragmentManager;
import ir.arapp.arappmain.Util.Services.HideShowKeyboard;
import ir.arapp.arappmain.Util.Services.SnackBarMessage;
import ir.arapp.arappmain.Util.Services.SnackBarToast;
import ir.arapp.arappmain.databinding.FragmentPhoneRegisterBinding;
import ir.arapp.arappmain.viewmodel.RegisterViewModel;

public class RegisterPhoneFragment extends Fragment implements SnackBarMessage, FragmentManager
{

//    region Variable
    FragmentPhoneRegisterBinding fragmentPhoneRegisterBinding;
    RegisterViewModel registerPhoneViewModel;
    private SnackBarToast snackBarToast;
    private HideShowKeyboard hideShowKeyboard;
//    endregion

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
//        Inflate the layout for this fragment
        fragmentPhoneRegisterBinding = FragmentPhoneRegisterBinding.inflate(inflater, container, false);
//        Set view model
        registerPhoneViewModel = ViewModelProviders.of(requireActivity()).get(RegisterViewModel.class);
        fragmentPhoneRegisterBinding.setViewModel(registerPhoneViewModel);
        registerPhoneViewModel.snackBarMessage = this;
        registerPhoneViewModel.fragmentManager = this;
//        Set Life Cycle
        fragmentPhoneRegisterBinding.setLifecycleOwner(this);
//        Hooks
        snackBarToast = new SnackBarToast(fragmentPhoneRegisterBinding.getRoot());
        hideShowKeyboard = new HideShowKeyboard(getContext(), fragmentPhoneRegisterBinding.getRoot());
//        Toolbar
        ((AppCompatActivity)requireActivity()).setSupportActionBar(fragmentPhoneRegisterBinding.registerToolbar);
        Objects.requireNonNull(((AppCompatActivity) requireActivity()).getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        fragmentPhoneRegisterBinding.registerToolbar.setNavigationOnClickListener(view1 -> onNavigateUp());
//        Return view
        return fragmentPhoneRegisterBinding.getRoot();
    }

//    region methods
//    On resume
    public void onResume()
    {
        super.onResume();
        requireActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                |  WindowManager.LayoutParams.FLAG_FULLSCREEN);
    }
//    Back button navigation
    private void onNavigateUp()
    {
        requireActivity().onBackPressed();
    }
//    Fragment navigation
    @Override
    public void navigateToFragment(String message)
    {
//        Nav Controller
        final NavController navController = Navigation.findNavController(requireView());

        switch (message)
        {
            case "login":
                navController.navigate(R.id.action_registerPhoneFragment_to_loginFragment);
                requireActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
                hideShowKeyboard.hideKeyboardFrom(true);
                break;
            case "validate":
                navController.navigate(R.id.action_registerPhoneFragment_to_registerValidateFragment);
                hideShowKeyboard.hideKeyboardFrom(true);
                break;
        }
    }
//    Set methods for specific function
    @Override
    public void setFunction(String type)
    {
    }
//    Error/success message
    @Override
    public void onSuccess(String message)
    {
        snackBarToast.snackBarShortTime(message);
    }
    @Override
    public void onFailure(String message)
    {
        snackBarToast.snackBarShortTime(message);
    }
//    endregion
}

File: app/src/main/java/ir/arapp/arappmain/view/fragments/RegisterPhoneFragment.kt
package ir.arapp.arappmain.view.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import ir.arapp.arappmain.R
import ir.arapp.arappmain.databinding.FragmentPhoneRegisterBinding
import ir.arapp.arappmain.util.services.FragmentManager
import ir.arapp.arappmain.util.services.HideShowKeyboard
import ir.arapp.arappmain.util.services.SnackBarMessage
import ir.arapp.arappmain.util.services.SnackBarToast
import ir.arapp.arappmain.viewmodel.RegisterViewModel
import java.util.*

class RegisterPhoneFragment : Fragment(), SnackBarMessage, FragmentManager {
    //    region Variable
    var _fragmentPhoneRegisterBinding: FragmentPhoneRegisterBinding? = null
    val fragmentPhoneRegisterBinding get() = _fragmentPhoneRegisterBinding!!
    private var snackBarToast: SnackBarToast? = null
    private var hideShowKeyboard: HideShowKeyboard? = null

    override fun onDestroyView() {
        super.onDestroyView()
        _fragmentPhoneRegisterBinding = null
    }
    //    endregion
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
//        Inflate the layout for this fragment
        _fragmentPhoneRegisterBinding =
            DataBindingUtil.inflate(inflater, R.layout.fragment_phone_register, container, false)
        //        Set view model
        val registerPhoneViewModel = ViewModelProvider(requireActivity()).get(
            RegisterViewModel::class.java
        )
        fragmentPhoneRegisterBinding.setViewModel(registerPhoneViewModel)
        registerPhoneViewModel.snackBarMessage = this
        registerPhoneViewModel.fragmentManager = this
        //        Set Life Cycle
        fragmentPhoneRegisterBinding.setLifecycleOwner(this)
        //        Hooks
        snackBarToast = SnackBarToast(fragmentPhoneRegisterBinding.getRoot())
        hideShowKeyboard = HideShowKeyboard(requireContext(), fragmentPhoneRegisterBinding.getRoot())
        //        Toolbar
        (requireActivity() as AppCompatActivity).setSupportActionBar(fragmentPhoneRegisterBinding.registerToolbar)
        Objects.requireNonNull((requireActivity() as AppCompatActivity).supportActionBar)!!
            .setDisplayHomeAsUpEnabled(true)
        fragmentPhoneRegisterBinding.registerToolbar.setNavigationOnClickListener { view1: View? -> onNavigateUp() }
        //        Return view
        return fragmentPhoneRegisterBinding.getRoot()
    }

    //    region methods
    //    On resume
    override fun onResume() {
        super.onResume()
        requireActivity().window.clearFlags(
            WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                    or WindowManager.LayoutParams.FLAG_FULLSCREEN
        )
    }

    //    Back button navigation
    private fun onNavigateUp() {
        requireActivity().onBackPressed()
    }

    //    Fragment navigation
    override fun navigateToFragment(message: String?) {
//        Nav Controller
        val navController = Navigation.findNavController(requireView())
        when (message) {
            "login" -> {
                navController.navigate(R.id.action_registerPhoneFragment_to_loginFragment)
                requireActivity().window.addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
                hideShowKeyboard!!.hideKeyboardFrom(true)
            }
            "validate" -> {
                navController.navigate(R.id.action_registerPhoneFragment_to_registerValidateFragment)
                hideShowKeyboard!!.hideKeyboardFrom(true)
            }
        }
    }

    //    Set methods for specific function
    override fun setFunction(type: String?) {}

    //    Error/success message
    override fun onSuccess(message: String?) {
        snackBarToast!!.snackBarShortTime(message, fragmentPhoneRegisterBinding!!.loginPanelLayout)
    }

    override fun onFailure(message: String?) {
        snackBarToast!!.snackBarShortTime(message, fragmentPhoneRegisterBinding!!.loginPanelLayout)
    } //    endregion
}